File: src/sqa-automation/level-0-beginner/golang/calculator_test.go
// Level 0 - Beginner: First Go Test
// ==================================
// Go ships with a built-in testing package — no extra dependencies needed.
//
// Run:
//   go test -v ./...        (from the golang/ directory)
//   go test -v -run TestCalculator

package level0

import (
	"errors"
	"testing"
)

// ── System Under Test ─────────────────────────────────────────────────────────

// Calculator is a simple arithmetic helper used as the SUT.
type Calculator struct{}

func (c Calculator) Add(a, b float64) float64      { return a + b }
func (c Calculator) Subtract(a, b float64) float64 { return a - b }
func (c Calculator) Multiply(a, b float64) float64 { return a * b }

func (c Calculator) Divide(a, b float64) (float64, error) {
	if b == 0 {
		return 0, errors.New("cannot divide by zero")
	}
	return a / b, nil
}

// ── Tests ─────────────────────────────────────────────────────────────────────

// Go test functions must start with "Test" and accept *testing.T.

func TestCalculatorAdd(t *testing.T) {
	calc := Calculator{}
	got := calc.Add(3, 4)
	want := float64(7)
	if got != want {
		t.Errorf("Add(3, 4) = %v; want %v", got, want)
	}
}

func TestCalculatorSubtract(t *testing.T) {
	calc := Calculator{}
	got := calc.Subtract(10, 3)
	if got != 7 {
		t.Errorf("Subtract(10, 3) = %v; want 7", got)
	}
}

func TestCalculatorMultiply(t *testing.T) {
	calc := Calculator{}
	if calc.Multiply(6, 7) != 42 {
		t.Error("Multiply(6, 7) should equal 42")
	}
}

func TestCalculatorDivide(t *testing.T) {
	calc := Calculator{}
	got, err := calc.Divide(10, 4)
	if err != nil {
		t.Fatalf("unexpected error: %v", err)
	}
	if got != 2.5 {
		t.Errorf("Divide(10, 4) = %v; want 2.5", got)
	}
}

func TestCalculatorDivideByZeroReturnsError(t *testing.T) {
	calc := Calculator{}
	_, err := calc.Divide(5, 0)
	if err == nil {
		t.Error("expected error when dividing by zero, got nil")
	}
}

func TestCalculatorAddNegativeNumbers(t *testing.T) {
	calc := Calculator{}
	if calc.Add(-3, -4) != -7 {
		t.Error("Add(-3, -4) should equal -7")
	}
}

func TestCalculatorAddZero(t *testing.T) {
	calc := Calculator{}
	if calc.Add(0, 42) != 42 {
		t.Error("Add(0, 42) should equal 42")
	}
}

// ── Table-driven tests ────────────────────────────────────────────────────────
// Table-driven tests are idiomatic Go — define a slice of test cases,
// then iterate, calling t.Run() for each one.

func TestCalculatorAddTableDriven(t *testing.T) {
	tests := []struct {
		name string
		a, b float64
		want float64
	}{
		{"positive + positive", 3, 4, 7},
		{"negative + negative", -3, -4, -7},
		{"zero identity", 0, 42, 42},
		{"float values", 1.5, 2.5, 4.0},
	}

	calc := Calculator{}
	for _, tc := range tests {
		t.Run(tc.name, func(t *testing.T) {
			got := calc.Add(tc.a, tc.b)
			if got != tc.want {
				t.Errorf("Add(%v, %v) = %v; want %v", tc.a, tc.b, got, tc.want)
			}
		})
	}
}

File: src/sqa-automation/level-0-beginner/java/CalculatorTest.java
// Level 0 - Beginner: First JUnit 5 Test in Java
// ================================================
// This file demonstrates a minimal JUnit 5 test for a Calculator class.
//
// Prerequisites:
//   JDK 17+, Maven or Gradle
//   Maven dependency: org.junit.jupiter:junit-jupiter:5.10.x
//
// Run:
//   mvn test   (from the java/ directory containing pom.xml)

package sqa.level0;

import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import static org.junit.jupiter.api.Assertions.*;

/**
 * Unit tests for the Calculator class.
 *
 * JUnit 5 discovers methods annotated with @Test automatically.
 * @DisplayName provides a human-readable description in test reports.
 */
@DisplayName("Calculator")
class CalculatorTest {

    // ── System Under Test ────────────────────────────────────────────────────

    /**
     * A simple calculator — our System Under Test (SUT).
     * In real projects this would live in src/main/java.
     */
    static class Calculator {
        double add(double a, double b)      { return a + b; }
        double subtract(double a, double b) { return a - b; }
        double multiply(double a, double b) { return a * b; }

        double divide(double a, double b) {
            if (b == 0) throw new ArithmeticException("Cannot divide by zero");
            return a / b;
        }
    }

    private final Calculator calc = new Calculator();

    // ── Tests ─────────────────────────────────────────────────────────────────

    @Test
    @DisplayName("adds two positive numbers")
    void addTwoPositiveNumbers() {
        double result = calc.add(3, 4);
        assertEquals(7, result);
    }

    @Test
    @DisplayName("subtracts to give correct difference")
    void subtractReturnsCorrectDifference() {
        double result = calc.subtract(10, 3);
        assertEquals(7, result);
    }

    @Test
    @DisplayName("multiplies two numbers")
    void multiplyTwoNumbers() {
        assertEquals(42, calc.multiply(6, 7));
    }

    @Test
    @DisplayName("divides to give a float result")
    void divideGivesFloatResult() {
        assertEquals(2.5, calc.divide(10, 4));
    }

    @Test
    @DisplayName("throws ArithmeticException when dividing by zero")
    void divideByZeroThrowsException() {
        ArithmeticException ex = assertThrows(
                ArithmeticException.class,
                () -> calc.divide(5, 0)
        );
        assertEquals("Cannot divide by zero", ex.getMessage());
    }

    @Test
    @DisplayName("adding negative numbers returns negative sum")
    void addNegativeNumbers() {
        assertEquals(-7, calc.add(-3, -4));
    }

    @Test
    @DisplayName("adding zero returns the other number")
    void addZeroReturnsOtherNumber() {
        assertEquals(42, calc.add(0, 42));
    }
}
